dask/bag/core.py

```diff
@@ -1507,6 +1582,186 @@ def bag_zip(*bags):
+
+
+def unpack_scalar_dask_kwargs(kwargs):
+    """Extracts dask values from kwargs
+
+    Currently only ``dask.bag.Item`` and ``dask.delayed.Delayed`` are
+    supported.  Returns a merged dask graph and a task resulting in a keyword
+    dict.
+    """
+    dsk = {}
+    kwargs2 = {}
+    for k, v in kwargs.items():
+        if isinstance(v, (Delayed, Item)):
+            dsk.update(v.dask)
+            kwargs2[k] = v.key
+        elif isinstance(v, Base):
+            raise NotImplementedError("dask.bag doesn't support kwargs of "
+                                      "type %s" % type(v).__name__)
+        else:
+            kwargs2[k] = v
+    if dsk:
+        kwargs = (dict, (zip, list(kwargs2), list(kwargs2.values())))
+    return dsk, kwargs
+
+
+def bag_map(func, *args, **kwargs):
+    """Apply a function elementwise across one or more bags.
+
+    Note that all ``Bag`` arguments must be partitioned identically.
+
+    Parameters
+    ----------
+    func : callable
+    *args, **kwargs : Bag, Item, Delayed, or object
+        Arguments and keyword arguments to pass to ``func``. Non-Bag args/kwargs
+        are broadcasted across all calls to ``func``.
+
+    Notes
+    -----
+    For calls with multiple `Bag` arguments, corresponding partitions should
+    have the same length; if they do not, the "extra" elements from the longer
+    partition(s) will be dropped.
+
+    Examples
+    --------
+    >>> import dask.bag as db
+    >>> b = db.from_sequence(range(5), npartitions=2)
+    >>> b2 = db.from_sequence(range(5, 10), npartitions=2)
+
+    Apply a function to all elements in a bag:
+
+    >>> db.map(lambda x: x + 1, b).compute()
+    [1, 2, 3, 4, 5]
+
+    Apply a function with arguments from multiple bags:
+
+    >>> from operator import add
+    >>> db.map(add, b, b2).compute()
+    [5, 7, 9, 11, 13]
+
+    Non-bag arguments are broadcast across all calls to the mapped function:
+
+    >>> db.map(add, b, 1).compute()
+    [1, 2, 3, 4, 5]
+
+    Keyword arguments are also supported, and have the same semantics as
+    regular arguments:
+
+    >>> def myadd(x, y=0):
+    ...     return x + y
+    >>> db.map(myadd, b, y=b2).compute()
+    [5, 7, 9, 11, 13]
+    >>> db.map(myadd, b, y=1).compute()
+    [1, 2, 3, 4, 5]
+
+    Both arguments and keyword arguments can also be instances of
+    ``dask.bag.Item`` or ``dask.delayed.Delayed``. Here we'll add the max value
+    in the bag to each element:
+
+    >>> db.map(myadd, b, b.max()).compute()
+    [4, 5, 6, 7, 8]
+    """
+    name = 'map-%s-%s' % (funcname(func), tokenize(func, args, kwargs))
+    dsk = {}
+
+    bags = []
+    args2 = []
+    for a in args:
+        if isinstance(a, Bag):
+            bags.append(a)
+            args2.append(a)
+            dsk.update(a.dask)
+        elif isinstance(a, (Item, Delayed)):
+            args2.append((itertools.repeat, a.key))
+            dsk.update(a.dask)
+        else:
+            args2.append((itertools.repeat, a))
```

Comment:
Agreed.  Just bringing it up.  I think that an alternative approach would be to look for a `.dask` and `.key` attribute or something.  This would be a new convention though.  I'm generally +1 on this PR as is.